common/devdoc/gballoc_design.md

```diff
@@ -0,0 +1,40 @@
+gballoc design
+=====
+
+`gballoc` is a layer that abstrats memory allocators. The need to have and be able to compare multiple memory allocators stems from performance requirements. Not all memory allocators have the same performance and for the sake of being able to compare them gballoc exists.
+
+`gballoc` is dvided in 2 layers (`ll` stand for `lower layer`, `hl` stands for `higher layer`):
+a) `gballoc_ll` - contains software wrappers over the memory allocation as provided by other components. For example, `gballoc_malloc/free` will redirect to `HeapAlloc/Free` when Windows APIs are directly used.
+b) `gballoc_hl` - contains performance measurements for `gballoc_ll` and it is build on top of `gballoc_ll`.
+
+
+`gballoc_ll` consists of a header file (`gballoc_ll.h`) and several possible implementation of its functions (i.e.: `gballoc_ll_crt.c`, `gballoc_ll_win.c`, `gballoc_ll_mimalloc.c` etc).
+
+`gballoc_hl` consists of a header file (`gballoc_hl.h`) and several possible implementation of the performance functions(i.e. `gballoc_hl_passthrough.c`, `gballoc_hl_buckets.c` etc)
+
+Both `gballoc_ll` and `gballoc_hl` have their own set of orthogonal configurations in CMake. That is - it is possible to use any combination of `gballoc_ll` implementation with any other implementation of `gballoc_ll`. 
+
+As far as Windows is concerned there are several CMake options.
+
+  1. `USE_SEGMENTED_HEAP` - this is a process wide option that replaces all the "regular" heaps with segmented heaps (https://docs.microsoft.com/en-us/windows/win32/sbscs/application-manifests#heaptype)
+
+  2. One a heap type is used, there are other several CMakeLists switches that further influence the SW behavior.
+
+    a) `gballoc_ll` implementation is switched between its implementations by the CMake option `GBALLOC_LL_TYPE` (of type string) which can be either
+
+      i. "PASSTHROUGH" - `gballoc_ll` shall route its calls directly to C's routines.
```

Comment:
>PASSTHROUGH [](start = 10, length = 11)

NIT can we have this as CRT? there is also a `PASSTHROUGH` below with a totally different meaning ...

Reply:
I would keep it "as is" APIs are called gballoc_ll_bla_bla_bla which calls (by passthrough) bla_bla_bla. 

And the HL layer's APIs of gballoc_hl_bla_bla_bla call (when passthrough) gballoc_ll_bla_bla_bla (lower layer).

---
In reply to: [457676593](https://github.com/Azure/azure-c-pal/pull/11#discussion_r457676593) [](ancestors = 457676593)